testing/unittests/jthreadtests.cpp

```diff
@@ -0,0 +1,591 @@
+        CPPUNIT_ASSERT_EQUAL(3U, threadCompletedCount.load());
+        CPPUNIT_ASSERT_EQUAL(0U, pool->runningCount());
+        CPPUNIT_ASSERT(pool->running());
+    }
+
+    void testThrottledThreadPool()
+    {
+        // Test: Expected throttling delays are introduced when pool is at capacity
+        pool.setown(createThreadPool("ThrottledTestPool", factory, true, nullptr, 3, 500));
+
+        auto startTime = std::chrono::high_resolution_clock::now();
+
+        unsigned lifespan2seconds = 2;
+        pool->start(&lifespan2seconds, "Thread1");
+        pool->start(&lifespan2seconds, "Thread2");
+        pool->start(&lifespan2seconds, "Thread3");
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+
+        auto afterQuickStarts = std::chrono::high_resolution_clock::now();
+        auto quickStartDuration = std::chrono::duration_cast<std::chrono::milliseconds>(afterQuickStarts - startTime);
+
+        // These should start immediately without throttling
+        CPPUNIT_ASSERT(quickStartDuration.count() < 100);
+        CPPUNIT_ASSERT_EQUAL(3U, pool->runningCount());
+
+        // When pool is at capacity, startNoBlock should throw exception immediately
+        try
+        {
+            pool->startNoBlock(&lifespan2seconds);
+            CPPUNIT_FAIL("startNoBlock should have thrown an exception when pool is at capacity");
+        }
+        catch (IException *e)
+        {
+            StringBuffer msg;
+            e->errorMessage(msg);
+            CPPUNIT_ASSERT_EQUAL_STR("No threads available in pool ThrottledTestPool", msg.str());
+            e->Release();
+        }
+
+        auto throttleStartTime = std::chrono::high_resolution_clock::now();
+
+        // This should block for approximately 100ms (start delay) then throw exception
+        try
+        {
+            pool->start(&lifespan2seconds, "Thread4", 100); // timeout after 100ms
+            CPPUNIT_FAIL("Start should have thrown an exception as no slots would be available when start delay specified");
+        }
+        catch (IException *e)
+        {
+            auto throttleEndTime = std::chrono::high_resolution_clock::now();
+            auto throttleDuration = std::chrono::duration_cast<std::chrono::milliseconds>(throttleEndTime - throttleStartTime);
+
+            CPPUNIT_ASSERT(throttleDuration.count() >= 80 && throttleDuration.count() <= 150);
+
+            StringBuffer msg;
+            e->errorMessage(msg);
+            CPPUNIT_ASSERT_EQUAL_STR("No threads available in pool ThrottledTestPool", msg.str());
+            e->Release();
+        }
+
+        auto longThrottleStartTime = std::chrono::high_resolution_clock::now();
+
+        // This should block for the full 500ms delay then not throw exception
+        try
+        {
+            pool->start(&lifespan2seconds, "Thread4"); // Will use default delay of 500ms
+        }
+        catch (IException *e)
+        {
+            StringBuffer msg("Exception encountered: ");
```

Comment:
should this exception handling be same (common code) exception handling as the previous one? i.e. check message and time >= X